### ISSUE TITLE
Debugging performance is poor and can hang VS Code when spawning lots of isolates when there are breakpoints

### ISSUE THREAD

**Issue by orestesgaolin** [Dart-Code/Dart-Code#5178]
**Describe the bug**

After about 5-10 minutes the entire VS Code window freezes and I need to kill both code and Dart processes to continue working. Stopping the app via Xcode does not unlock VS Code.

My app can spawn couple of isolates when running (sometimes more than 10).

I have run flutter clean multiple times

**To Reproduce**

1. Run the app on iOS or Android device in debug mode for some time
2. Hit couple of breakpoints (5-8)
3. Edit few lines and hot reload
4. Observe that window completely freezes, cannot collect the diagnostics information etc. as window is non-responsive.

**Screenshots**

Error after which window is not responding

![screenshot_20240709_162234](https://github.com/Dart-Code/Dart-Code/assets/16854239/b883fe9e-2a94-4ea1-84ae-0fff214d3d09)

Isolates running 
![screenshot_20240709_161855](https://github.com/Dart-Code/Dart-Code/assets/16854239/d73adf2f-5b54-44ee-af5f-9569f152fd50)

**Please complete the following information:**

 - Operating System and version: macOS 14.5 (23F79)
 - VS Code version: Version: 1.91.0 (Universal)
 - Dart Code extension: 3.92.0
 - Flutter extension: 3.92.0 (activated)
 - Dart/Flutter SDK version: stable 3.22.2
 - Target device (if the issue relates to Flutter debugging): iPhone 12 mini running iOS 17.5

I have sent the report to [logs@dartcode.org](mailto:logs@dartcode.org) including
- analyzer logs
- instrumentation logs
- DAP logs
- extension startup log

**Comment by DanTup** [comment]
Thanks - from the log it looks like we're generating a large number of breakpoint change events:

```
[4:42:55 PM] [DAP] [Info] <== {"seq":282953,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282954,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282955,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282956,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282957,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282958,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282959,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282960,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282961,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282962,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282963,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282964,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282965,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282966,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282967,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282968,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282969,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282970,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282971,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282972,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282973,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282974,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282975,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282976,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282977,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282978,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282979,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
[4:42:55 PM] [DAP] [Info] <== {"seq":282980,"type":"event","body":{"breakpoint":{"column":64,"id":100000,"line":36,"verified":true},"reason":"changed"},"event":"breakpoint"}
```

Without the VM Service traffic it's hard to know if these are in response to real events from the VM though. Do you happen to know how many isolates are being spawned (and whether they're exiting and new ones being created, or if the entire number is small)?

I feel like VS Code _hanging_ is a VS Code bug, because we're running in another process and shouldn't be able to hang the editor at all (VS Code is usually good at protecting itself against any bad extension in another process), but it's possible we are sending these events faster than it can process them.

If you can find a way to repro this with a shareable project, that would significantly help track this down.

**Comment by DanTup** [comment]
I looked more closely at the log, and I can see that isolates are starting, doing work, then exiting. Each time this happens, there is 1 more isolate event - so I think we are definitely building up these events for each isolate that has been spawned.

I will try to reproduce this with an app and will post back.

**Comment by DanTup** [comment]
Oops, not sure how I closed this.

I can repro with this code:

```dart
import 'dart:isolate';

Future<void> main(List<String> arguments) async {
  // Breakpoint here, then start Capturing Logs, then resume.
  for (var i = 0; i < 50; i++) {
    await Isolate.run(() => foo(i));
  }
  await Future.delayed(const Duration(seconds: 2));
}

void foo(int i) {
  print(i);
  if (DateTime.now().year == 1) {
    print('test'); // Add a breakpoint here
  }
}
```

My log builds up the same events:

```
[11:43:00 AM] [DAP] [Info] <== {"seq":1970,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1971,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1972,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1973,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1974,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1975,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1976,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1977,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1978,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1979,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1980,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1981,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1982,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1983,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1984,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1985,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1986,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1987,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1988,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1989,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1990,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1991,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1992,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1993,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1994,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1995,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1996,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1997,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1998,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
[11:43:00 AM] [DAP] [Info] <== {"seq":1999,"type":"event","body":{"breakpoint":{"column":14,"id":100000,"line":4,"verified":true},"reason":"changed"},"event":"breakpoint"}
```

Each event also seems to trigger VS Code asking for threads:

```
[11:43:00 AM] [DAP] [Info] <== {"seq":2032,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":48,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":49}
[11:43:00 AM] [DAP] [Info] <== {"seq":2033,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":49,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":50}
[11:43:00 AM] [DAP] [Info] <== {"seq":2034,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":50,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":51}
[11:43:00 AM] [DAP] [Info] <== {"seq":2035,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":51,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":52}
[11:43:00 AM] [DAP] [Info] <== {"seq":2036,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":52,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":53}
[11:43:00 AM] [DAP] [Info] <== {"seq":2037,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":53,"success":true}
[11:43:00 AM] [DAP] [Info] ==> {"command":"threads","type":"request","seq":54}
[11:43:00 AM] [DAP] [Info] <== {"seq":2038,"type":"response","body":{"threads":[{"id":1,"name":"main"}]},"command":"threads","request_seq":54,"success":true}
```

It's probably bad behaviour of VS Code to not debounce/throttle these thread requests, but perhaps once we fix the original issue what's left will be so fast that it'll be insignificant.

**Comment by DanTup** [comment]
Actually, I can't reproduce this on the latest code. I think this might have been fixed by https://github.com/dart-lang/sdk/commit/5ed21576fecd39cb83811bd8191ac23c07e02490 which changed the way we handle breakpoints across isolates (better handling that breakpoint IDs can be reused by each isolate).

@orestesgaolin are you able to easily test a copy of your project with Flutter's `master` branch to confirm that this issue does not still exist for you there? (If it is still possible to trigger it, it'd be good to find out now rather than wait for that change to get to stable). Thanks!

**Comment by orestesgaolin** [comment]
I tested on beta, and the VS Code performance was better (not getting stuck) but the app would just stop after 10-20 minutes. I wasn't able to run in on master due to ndk issues, will try again tomorrow.

**Comment by DanTup** [comment]
The change noted above appears to already be on beta, so probably testing master won't change much.

Are you able to capture a new log on beta? I can't reproduce my issue on beta, but my repro was much simpler than yours. If you're strip your project down to something you can share that still reproduces the problem, that could also help debug. Thanks!

**Comment by orestesgaolin** [comment]
Just sent you the logs from the case where the app just froze when I wasn't doing anything special, isolates were stuck as above.

**Comment by DanTup** [comment]
Yes, beta logs are helpful here. However, it's hard to tell what's happening in this one because a) the log contains multiple debug sessions (and I don't know which is the bad one) and b) there's no VM Service traffic.

Is it possible you could capture logs using this command instead:

- Open the command palette (`F1`)
- Run the **Dart: Capture Debugging Logs** command
- Start your debug session
- Let the app crash/freeze
- Click **Cancel** on the logging notification to stop logging and open the log file

Logs captured this way (as long as you start logging before starting the debug session) will include the VM Service traffic (this will also make the performance worse, but it might at least show is what's going on).

If for any reason you need to start a new debug session to repro, please stop and restart logging, so the log is only for the problematic session. It may also be useful to note the time you started the debug session (and if possible, the time the app appeared to freeze).

**Comment by orestesgaolin** [comment]
Thanks, just sent the logs recorded with **Dart: Capture Debugging Logs** to your e-mail. The app got stuck just after hitting a breakpoint few seconds after launch.

**Comment by DanTup** [comment]
Sorry for the delay. I had a look through the log and nothing stands out.. the last events in the log show an isolate starting, breakpoints being sent to it, and then it resuming:

![image](https://github.com/user-attachments/assets/e42df847-a204-41ce-a55c-3fcad16c7857)

What might be interesting to see when the VM is in this state, is what the DevTools debugger shows. If you reproduce this, then run the **Open DevTools in Browser** command to get DevTools in a browser, remove `&hide=debugger` from the querystring in the URL if it's there, and then go to the Debugger page and see what isolates are listed and what the statuses of them are.

I'm also curious - is this a physical iPhone you're testing with? Does it occur the same on another device (macOS desktop or iOS simulator)?

**Comment by orestesgaolin** [comment]
Thanks, I will try to capture the VM state.

It happens no matter I use device or simulator.

I think I managed to find quite straightforward reproduction case on beta.

When I have a breakpoint set in the app, and I try to interact with the app after the breakpoint is hit(execution is paused), VS Code looses the information about the VM being in paused state -- none of the isolates is marked as paused anymore, but the app is not interactive. If I go to the Devtools I can see that the main isolate is still paused and resume it.


https://github.com/user-attachments/assets/bdc6e9b1-cd4f-4c88-8ef1-46ab1c5d16f3

**Comment by DanTup** [comment]
It's not clear to me what's happening in the video - I see it hit the breakpoint and you press something off-screen, and then it looks like the app is still paused but VS Code isn't. What was it you did off-screen?

If you're able to repro this in something you can share, I'd really like to look at this, it's quite concerning. Does it only occur with lots of isolates? The isolates make the log file more difficult to read, but a log file without all the extra isolates with this happening might be easier to examine (a repro I can run is ofc better though, because I can iterate/debug much faster).

**Comment by DanTup** [comment]
> on beta, I cannot even open the devtools in browser, they are stuck on loading screen

Hmm, that's odd. Is this Chrome? (the address bar/icons look different to me). If not, does using Chrome make any difference?

Are there any errors in the developer console?

If you can make a repro, that would definitely help debug.

**Comment by DanTup** [comment]
@orestesgaolin is this still an issue? Were you able to make an isolated repro?

**Comment by mraleph** [comment]
@orestesgaolin says that he is still experiencing the issue all the time. I wonder what we can do to narrow it down if we can't obtain the repro.

**Comment by orestesgaolin** [comment]
The exact issue is varying in behavior. Sometimes it's exactly as explained in the issue where the VS Code freezes, but it's fairly rare now. I need to create new ticket with the current behavior which is slightly different:
- app becomes unresponsive, no touch events are handled
- autocomplete becomes unusable, symbols cannot be found
- app sometimes crashes when paused on a breakpoint
- hot restart does not complete

To reproduce it I need to run the app in debug mode for couple of minutes, do few dozens of hot reloads etc.

**Comment by DanTup** [comment]
I think opening separate issues for the different problems would be best, as each of the things above sounds like it could be a distinct issue (and it's easier to close dupes than separate them out).

If you're unable to build a repro that can be shared, the debugging logs (as discussed above) would be useful - especially so paired with a screen recording and/or description of exactly what's going on (a visible clock to help match up the logs to what's happening would also help if the steps include the same steps being repeated multiple times).

There will always be some additional latency when debugging with isolates, because we must set each isolated to start paused (so that we can send breakpoints before any code runs, set the exception-pause preferences, etc.). The debug adapter is single threaded, so if a large number of isolates spawn, all of the code to send these API calls (and then resume the isolate) happens in a single thread. It's hard to say if what you're seeing is related to this or not though.

(Conditional breakpoints in frequently hit code can slow things down a lot, because they are effectively normal breakpoints that are hit, and then the expression is evaluated by the debug adapter and the response determines whether to automatically resume or stay paused).
